lasagne/layers/recurrent.py

```diff
@@ -0,0 +1,1239 @@
+by using a few reshape operations; please refer to the example below.
+
+Examples
+--------
+The following example demonstrates how recurrent layers can be easily mixed
+with feed-forward layers using :class:`ReshapeLayer` and how to build a
+network with variable batch size and number of time steps.
+
+>>> from lasagne.layers import *
+>>> num_inputs, num_units, num_classes = 10, 12, 5
+>>> # By setting the first two dimensions as None, we are allowing them to vary
+>>> # They correspond to batch size and sequence length, so we will be able to
+>>> # feed in batches of varying size with sequences of varying length.
+>>> l_inp = InputLayer((None, None, num_inputs))
+>>> # We can retrieve symbolic references to the input variable's shape, which
+>>> # we will later use in reshape layers.
+>>> batchsize, seqlen, _ = l_inp.input_var.shape
+>>> l_lstm = LSTMLayer(l_inp, num_units=num_units)
+>>> # In order to connect a recurrent layer to a dense layer, we need to
+>>> # flatten the first two dimensions (our "sample dimensions"); this will
+>>> # cause each time step of each sequence to be processed independently
+>>> l_shp = ReshapeLayer(l_lstm, (-1, num_units))
+>>> l_dense = DenseLayer(l_shp, num_units=num_classes)
+>>> # To reshape back to our original shape, we can use the symbolic shape
+>>> # variables we retrieved above.
+>>> l_out = ReshapeLayer(l_dense, (batchsize, seqlen, num_classes))
+"""
+import numpy as np
+import theano
+import theano.tensor as T
+from .. import nonlinearities
+from .. import init
+from ..utils import unroll_scan
+
+from .base import Layer
+from .input import InputLayer
+from .dense import DenseLayer
+from . import helper
+
+__all__ = [
+    "CustomRecurrentLayer",
+    "RecurrentLayer",
+    "Gate",
+    "LSTMLayer",
+    "GRULayer"
+]
+
+
+class CustomRecurrentLayer(Layer):
+    """
+    lasagne.layers.recurrent.CustomRecurrentLayer(incoming, input_to_hidden,
+    hidden_to_hidden, nonlinearity=lasagne.nonlinearities.rectify,
+    hid_init=lasagne.init.Constant(0.), backwards=False,
+    learn_init=False, gradient_steps=-1, grad_clipping=False,
+    unroll_scan=False, precompute_input=True, **kwargs)
+
+    A layer which implements a recurrent connection.
+
+    This layer allows you to specify custom input-to-hidden and
+    hidden-to-hidden connections by instantiating layer instances and passing
+    them on initialization.  The output shape for the provided layers must be
+    the same.  If you are looking for a standard, densely-connected recurrent
+    layer, please see :class:`RecurrentLayer`.  The output is computed
+    by
+
+    .. math ::
+        h_t = \sigma(f_i(x_t) + f_h(h_{t-1}))
+
+    Parameters
+    ----------
+    incoming : a :class:`lasagne.layers.Layer` instance or a tuple
+        The layer feeding into this layer, or the expected input shape.
+    input_to_hidden : :class:`lasagne.layers.Layer`
+        Layer which connects input to the hidden state (:math:`f_i`).
```

Comment:
> It doesn't matter if it's a layer created from shape,

It can't be a layer created with a tuple as `incoming`, because e.g. at https://github.com/craffel/nntools/blob/recurrent/lasagne/layers/recurrent.py#L277 `get_output` is called, which will throw an error:

```Python
In [1]: import lasagne
In [2]: l_in = lasagne.layers.InputLayer((2, 3, 4))
In [3]: l_rec = lasagne.layers.CustomRecurrentLayer(l_in, lasagne.layers.DenseLayer((2, 4), 6), lasagne.layers.DenseLayer((2, 6), 6))
In [4]: out = lasagne.layers.get_output(l_rec)
---------------------------------------------------------------------------
ValueError                                Traceback (most recent call last)
<ipython-input-4-0df317cbaa26> in <module>()
----> 1 out = lasagne.layers.get_output(l_rec)

lasagne/layers/helper.pyc in get_output(layer_or_layers, inputs, **kwargs)
    226                                  "mapping this layer to an input expression."
    227                                  % layer)
--> 228             all_outputs[layer] = layer.get_output_for(layer_inputs, **kwargs)
    229     # return the output(s) of the requested layer(s) only
    230     try:

lasagne/layers/recurrent.pyc in get_output_for(self, input, mask, **kwargs)
    275                               (seq_len*num_batch, -1))
    276             input = helper.get_output(
--> 277                 self.input_to_hidden, input, **kwargs)
    278
    279             # Reshape back to (seq_len, batch_size, num_units)

lasagne/layers/helper.pyc in get_output(layer_or_layers, inputs, **kwargs)
    225                                  "layer %r. Please call it with a dictionary "
    226                                  "mapping this layer to an input expression."
--> 227                                  % layer)
    228             all_outputs[layer] = layer.get_output_for(layer_inputs, **kwargs)
    229     # return the output(s) of the requested layer(s) only

ValueError: get_output() was called without giving an input expression for the free-floating layer <lasagne.layers.dense.DenseLayer object at 0x7f057952ee10>. Please call it with a dictionary mapping this layer to an input expression.
```

> It's just that the documentation currently reads as if it should be a single layer, not a network (although that's technically just a layer instance as well, of course).

Where does it sound like that?

```
This layer allows you to specify custom input-to-hidden and
hidden-to-hidden connections by instantiating layer instances and passing
them on initialization.
```
?

That's meant to read that you can instantiate multiple layers for a input-to-hidden connection and multiple layers for a hidden-to-hidden connection.